simphony/engine/extension.py

```diff
@@ -0,0 +1,219 @@
+          supported engine interfaces as a list of engine.EngineInterface enums
+
+        Returns
+        -------
+        EngineMetadata: a loaded engine metadata object
+        """
+        return EngineMetadata(name, features, interfaces)
+
+    @staticmethod
+    def create_engine_metadata_feature(physics_equation, methods):
+        """Factory method to create engine metadata feature.
+
+        Parameters
+        ----------
+        physics_equation: PhysicsEquation
+          represents a physics equation
+        methods: list
+          ComputationalMethod objects to solve the equation
+
+        Return
+        ------
+        EngineFeatureMetadata: a loaded engine metadata feature object
+        """
+        return EngineFeatureMetadata(physics_equation, methods)
+
+
+class EngineManager(object):
+    """Controller class to keep track of engine extensions."""
+    def __init__(self):
+        self._engine_extensions = {}
+
+    def load_metadata(self, plugin):
+        """Load existing engine extensions into the manager.
+
+        This method inspects the given module and looks for ABCEngineExtension
+         subclasses. If provided, will instantiate and keep them for further
+         interactions with corresponding engines.
+
+        plugin: module
+          A python module provided by an extension
+        """
+        if not inspect.ismodule(plugin):
+            raise EngineManagerException(
+                'The provided object is not a module: %s' % plugin)
+
+        for name, value in inspect.getmembers(plugin, inspect.isclass):
+            if not inspect.isabstract(value) and\
+                    issubclass(value, ABCEngineExtension):
+                # Load the corresponding engine extension
+                extension = value()
+                if not isinstance(extension, ABCEngineExtension):
+                    raise ValueError('Expected ABCEngineExtension, got %s' %
+                                     extension)
+                self.add_extension(extension)
+
+    def add_extension(self, extension):
+        """Add an extension to the context.
+
+        Parameters
+        ----------
+        extension: engine.ABCEngineExtension
+          an extension that has knowledge about its own engines.
+        """
+        for engine in extension.get_supported_engines():
+            if engine.name in self._engine_extensions:
+                raise Exception('There is already an extension registered'
+                                ' for %s engine.' % engine.name)
+
+            # This extension knows how to create wrappers for the given engine.
+            # It is our interaction point with its engines.
+            self._engine_extensions[engine.name] = extension
+
+    def create_wrapper(self, cuds, engine_name, engine_interface=None):
+        """Create a wrapper to the given engine.
+
+        Parameters
+        ----------
+        cuds: CUDS
+            A cuds object which contains model information.
```

Comment:
Again probably better to change these description lines to not have a capital first letter and a full stop at the end.